src/java.base/share/classes/java/util/LazyCollections.java

```diff
@@ -0,0 +1,650 @@
+                return new LazySubList<>(list, fromIndex, toIndex - fromIndex);
+            }
+
+            static <E> ImmutableCollections.SubList<E> fromLazySubList(LazySubList<E> parent, int fromIndex, int toIndex) {
+                return new LazySubList<>(parent.root, parent.offset + fromIndex, toIndex - fromIndex);
+            }
+
+        }
+
+        private static final class ReverseOrderLazyListView<E>
+                extends ReverseOrderListView.Rand<E> {
+
+            private ReverseOrderLazyListView(List<E> base) {
+                super(base, false);
+            }
+
+            @Override
+            public List<E> subList(int fromIndex, int toIndex) {
+                final int size = base.size();
+                subListRangeCheck(fromIndex, toIndex, size);
+                return new ReverseOrderLazyListView<>(base.subList(size - toIndex, size - fromIndex));
+            }
+        }
+    }
+
+    static final class LazyEnumMap<K extends Enum<K>, V>
+            extends AbstractLazyMap<K, V> {
+
+        @Stable
+        private final Class<K> enumType;
+        @Stable
+        // We are using a wrapper class here to be able to use a min value of zero that
+        // is also stable.
+        private final Integer min;
+        @Stable
+        private final IntPredicate member;
+
+        public LazyEnumMap(Set<K> set,
+                           Class<K> enumType,
+                           int min,
+                           int backingSize,
+                           IntPredicate member,
+                           Function<? super K, ? extends V> computingFunction) {
+            this.enumType = enumType;
+            this.min = min;
+            this.member = member;
+            super(set, set.size(), backingSize, computingFunction);
+        }
+
+        @Override
+        @ForceInline
+        public boolean containsKey(Object o) {
+            return enumType.isAssignableFrom(o.getClass())
+                    && member.test(((Enum<?>) o).ordinal());
+        }
+
+        @ForceInline
+        @Override
+        public V getOrDefault(Object key, V defaultValue) {
+            if (enumType.isAssignableFrom(key.getClass())) {
+                final int ordinal = ((Enum<?>) key).ordinal();
+                if (member.test(ordinal)) {
+                    @SuppressWarnings("unchecked")
+                    final K k = (K) key;
+                    return orElseCompute(k, indexForAsInt(k));
+                }
+            }
+            return defaultValue;
+        }
+
+        @Override
+        Integer indexFor(K key) {
+            return indexForAsInt(key);
+        }
+
+        private int indexForAsInt(K key) {
+            return key.ordinal() - min;
+        }
+
+    }
+
+    static final class LazyMap<K, V>
+            extends AbstractLazyMap<K, V> {
+
+        @Stable
+        private final Map<K, Integer> indexMapper;
```

Comment:
This index mapper is a clever solution that avoids implementing a hashing function again. Lookups through this map can be folded because it is created using `Map.ofEntries`. I think you should put that in a comment here as well.